File: discord/src/main/java/io/github/shaksternano/borgar/discord/media/graphics/drawable/ImageDrawable.java
package io.github.shaksternano.borgar.discord.media.graphics.drawable;

import io.github.shaksternano.borgar.discord.media.ImageFrame;
import io.github.shaksternano.borgar.discord.media.ImageUtil;
import io.github.shaksternano.borgar.discord.media.io.MediaReaders;
import io.github.shaksternano.borgar.discord.media.io.reader.MediaReader;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.InputStream;
import java.util.Objects;

public class ImageDrawable implements Drawable {

    private final MediaReader<ImageFrame> reader;
    private final BufferedImage firstFrame;
    private int targetWidth;
    private int targetHeight;
    private int actualWidth;
    private int actualHeight;

    public ImageDrawable(InputStream inputStream, String format) throws IOException {
        reader = MediaReaders.createImageReader(inputStream, format);
        targetWidth = reader.width();
        targetHeight = reader.height();
        firstFrame = reader.first().content();
        actualWidth = firstFrame.getWidth();
        actualHeight = firstFrame.getHeight();
    }

    @Override
    public void draw(Graphics2D graphics, int x, int y, long timestamp) throws IOException {
        var image = resizeImage(reader.readFrame(timestamp).content());
        graphics.drawImage(image, x, y, null);
    }

    private BufferedImage resizeImage(BufferedImage image) {
        image = ImageUtil.fitWidth(image, targetWidth);
        return ImageUtil.fitHeight(image, targetHeight);
    }

    @Override
    public int getWidth(Graphics2D graphicsContext) {
        return actualWidth;
    }

    @Override
    public int getHeight(Graphics2D graphicsContext) {
        return actualHeight;
    }

    @Override
    public Drawable resizeToWidth(int width) {
        if (width != actualWidth) {
            this.targetWidth = width;
            var resized = resizeImage(firstFrame);
            actualWidth = resized.getWidth();
            actualHeight = resized.getHeight();
        }
        return this;
    }

    @Override
    public Drawable resizeToHeight(int height) {
        if (height != actualHeight) {
            this.targetHeight = height;
            var resized = resizeImage(firstFrame);
            actualWidth = resized.getWidth();
            actualHeight = resized.getHeight();
        }
        return this;
    }

    @Override
    public int getFrameCount() {
        return reader.frameCount();
    }

    @Override
    public long getDuration() {
        return reader.duration();
    }

    @Override
    public boolean sameAsPreviousFrame() {
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(
            reader,
            targetWidth,
            targetHeight,
            actualWidth,
            actualHeight
        );
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        } else if (obj instanceof ImageDrawable other) {
            return Objects.equals(reader, other.reader)
                && targetWidth == other.targetWidth
                && targetHeight == other.targetHeight
                && actualWidth == other.actualWidth
                && actualHeight == other.actualHeight;
        } else {
            return false;
        }
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[reader=" + reader + ", width=" + targetWidth + ", height=" + targetHeight + "]";
    }

    @Override
    public void close() throws IOException {
        reader.close();
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/graphics/drawable/ImageDrawable.kt
package io.github.shaksternano.borgar.core.graphics.drawable

import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.io.fileFormat
import io.github.shaksternano.borgar.core.media.createImageReader
import io.github.shaksternano.borgar.core.media.reader.ImageReader
import io.github.shaksternano.borgar.core.media.reader.firstContent
import io.github.shaksternano.borgar.core.media.reader.readContent
import io.github.shaksternano.borgar.core.media.resizeHeight
import io.github.shaksternano.borgar.core.util.hash
import io.github.shaksternano.borgar.core.util.kClass
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import kotlin.time.Duration

class ImageDrawable internal constructor(
    private val dataSource: DataSource,
    private val reader: ImageReader,
    private val firstFrame: BufferedImage,
    private val width: Int = firstFrame.width,
    private val height: Int = firstFrame.height,
) : Drawable {

    override suspend fun draw(graphics: Graphics2D, x: Int, y: Int, timestamp: Duration) {
        val image = reader.readContent(timestamp).resizeHeight(height)
        graphics.drawImage(image, x, y, null)
    }

    override suspend fun getWidth(graphics: Graphics2D): Int = width

    override suspend fun getHeight(graphics: Graphics2D): Int = height

    override fun resizeToHeight(height: Int): Drawable {
        return if (height == this.height) {
            this
        } else {
            val resizedImage = firstFrame.resizeHeight(height)
            ImageDrawable(
                dataSource,
                reader,
                firstFrame,
                resizedImage.width,
                height,
            )
        }
    }

    override suspend fun close() = reader.close()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (kClass != other?.kClass) return false

        other as ImageDrawable

        if (dataSource != other.dataSource) return false
        if (width != other.width) return false
        if (height != other.height) return false

        return true
    }

    override fun hashCode(): Int = hash(
        dataSource,
        width,
        height,
    )

    override fun toString(): String {
        return "ImageDrawable(dataSource=$dataSource" +
            ", reader=$reader" +
            ", firstFrame=$firstFrame" +
            ", width=$width" +
            ", height=$height)"
    }
}

suspend fun ImageDrawable(
    input: DataSource,
): ImageDrawable {
    val reader = createImageReader(input, input.fileFormat())
    return ImageDrawable(
        input,
        reader,
        reader.firstContent(),
    )
}
